src/main/java/uk/gov/hmcts/dts/fact/services/search/PostcodeSearchForServiceAreaRunner.java

```diff
@@ -0,0 +1,105 @@
+import uk.gov.hmcts.dts.fact.repositories.CourtWithDistanceRepository;
+
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.stream.Collectors.toList;
+import static uk.gov.hmcts.dts.fact.model.ServiceAreaType.CIVIL;
+import static uk.gov.hmcts.dts.fact.model.ServiceAreaType.FAMILY;
+
+@Component
+public class PostcodeSearchForServiceAreaRunner {
+
+    private static final String LOCAL_AUTHORITY = "local-authority";
+    private static final String REGIONAL = "regional";
+
+    private final CourtWithDistanceRepository courtWithDistanceRepository;
+
+    public PostcodeSearchForServiceAreaRunner(CourtWithDistanceRepository courtWithDistanceRepository) {
+        this.courtWithDistanceRepository = courtWithDistanceRepository;
+    }
+
+    public List<CourtReferenceWithDistance> getSearchFor(final ServiceArea serviceArea,
+                                                         final String postcode,
+                                                         final MapitData mapitData) {
+
+        final String type = serviceArea.getType();
+        final String areaOfLaw = serviceArea.getAreaOfLaw().getName();
+        final Double lat = mapitData.getLat();
+        final Double lon = mapitData.getLon();
+
+        if (FAMILY.isEqualTo(type)) {
+
+            final Optional<String> localAuthorityOptional = mapitData.getLocalAuthority();
+            
+            if (LOCAL_AUTHORITY.equals(serviceArea.getCatchmentMethod())
+                && localAuthorityOptional.isPresent()) {
+
+                final String localAuthority = localAuthorityOptional.get();
+
+                if (isRegional(serviceArea)) {
+
+                    final List<CourtWithDistance> courtsWithDistance = courtWithDistanceRepository
+                        .findNearestRegionalByAreaOfLawAndLocalAuthority(lat, lon, areaOfLaw, localAuthority);
+
+                    return fallbackProximityRegionalSearchIfEmpty(courtsWithDistance, areaOfLaw, mapitData);
+                } else {
+
+                    final List<CourtWithDistance> courtsWithDistance = courtWithDistanceRepository
+                        .findNearestTenByAreaOfLawAndLocalAuthority(lat, lon, areaOfLaw, localAuthority);
+
+                    return fallbackProximitySearchIfEmpty(courtsWithDistance, areaOfLaw, mapitData);
+                }
+            }
+
+        } else if (CIVIL.isEqualTo(type)) {
+
+            final List<CourtWithDistance> courtsWithDistance = courtWithDistanceRepository
+                .findNearestTenByAreaOfLawAndCourtPostcode(lat, lon, areaOfLaw, postcode);
+
+            return fallbackProximitySearchIfEmpty(courtsWithDistance, areaOfLaw, mapitData);
+        }
+
+        return convert(courtWithDistanceRepository.findNearestTenByAreaOfLaw(lat, lon, areaOfLaw));
+    }
+
+    private boolean isRegional(final ServiceArea serviceArea) {
+        return serviceArea.getServiceAreaCourts().stream()
```

Comment:
perhaps this could be a method on the service area model?

Reply:
Moved into ServiceArea class